molgenis-data/src/main/java/org/molgenis/data/listeners/EntityListenersService.java

```diff
@@ -0,0 +1,140 @@
+package org.molgenis.data.listeners;
+
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.SetMultimap;
+import org.molgenis.data.Entity;
+import org.molgenis.data.MolgenisDataException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Service;
+
+import javax.annotation.PreDestroy;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Stream;
+
+import static java.util.Objects.requireNonNull;
+
+@Service
+public final class EntityListenersService
+{
+	private final Logger LOG = LoggerFactory.getLogger(EntityListenersService.class);
+	private final ConcurrentHashMap<String, SetMultimap<Object, EntityListener>> entityListeners = new ConcurrentHashMap<>();
+
+	/**
+	 * Register a repository to the entity listeners service once
+	 *
+	 * @param repoFullName
+	 * @return
+	 */
+	void register(String repoFullName)
+	{
+		if (!entityListeners.containsKey(requireNonNull(repoFullName)))
+		{
+			entityListeners.put(repoFullName, HashMultimap.create());
+		}
+	}
+
+	/**
+	 * Update all registered listeners of the entities
+	 *
+	 * @param repoFullName
+	 * @param entities
+	 * @return Stream<Entity>
+	 */
+	Stream<Entity> updateEntities(String repoFullName, Stream<Entity> entities)
+	{
+		verifyRepoRegistered(repoFullName);
+		SetMultimap<Object, EntityListener> entityListeners = this.entityListeners.get(repoFullName);
+		return entities.filter(entity -> {
+			Set<EntityListener> entityEntityListeners = entityListeners.get(entity.getIdValue());
+			entityEntityListeners.forEach(entityListener -> {
+				entityListener.postUpdate(entity);
```

Comment:
After some googling: How about we use a http://tutorials.jenkov.com/java-util-concurrent/readwritelock.html then we can let all read access to the listeners happen in a block that does lock.readLock().lock() and inserting or removing listeners happen in a block that does lock.writeLock().lock() !